dali/kernels/shape.h

```diff
@@ -0,0 +1,299 @@
+// ptrdiff_t CalcOffset(const Shape &shape, const SamplePos &pos) {
+//   ptrdiff_t ofs = pos[0];
+//   const int m = ShapeDim(pos);
+//   const int n = ShapeDim(shape);
+//   int i;
+//   for (i = 1; i < m; i++) {
+//     ofs *= shape[i];
+//     ofs += pos[i];
+//   }
+//   for (; i < n; i++) {
+//     ofs *= shape[i];
+//   }
+//   return ofs;
+// }
+
+// /// @brief Returns the product of all elements in shapeZ
+// /// @param shape - shape of a tensor whose elements we count
+// template <typename Shape>
+// inline int64_t Volume(const Shape &shape) {
+//   int n = ShapeDim(shape);
+//   if (n < 1)
+//     return 0;
+//   int64_t v = shape[0];
+//   for (int i = 1; i < n; i++) {
+//     v *= shape[i];
+//   }
+//   return v;
+// }
+
+constexpr int DynamicTensorShape = -1;
+
+template <int dim_>
+struct TensorShape;
+
+template <>
+struct TensorShape<DynamicTensorShape> {
+  TensorShape(const std::vector<int64_t> &s) : shape(s) {}
+  TensorShape(std::vector<int64_t> &&s) : shape(std::move(s)) {}
+  template <size_t N>
+  TensorShape(std::array<int64_t, N> s) : shape(s.begin(), s.end()) {}
+
+  TensorShape() = default;
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  // TODO(klecki) efficient size calculation for this vector?
+  template <int other_dim>
+  TensorShape(const TensorShape<other_dim> &other) : shape(other.shape.begin(), other.shape.end()) {
+    static_assert(other_dim != DynamicTensorShape, "This constructor should be not used");
+  }
+
+  template <int other_dim>
+  TensorShape(TensorShape<other_dim> &&other) : shape(other.shape.begin(), other.shape.end()) {}
+
+  template <int other_dim>
+  TensorShape &operator=(const TensorShape<other_dim> &other) {
+    shape = std::vector<int64_t>(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  template <int other_dim>
+  TensorShape &operator=(TensorShape<other_dim> &&other) {
+    shape = std::vector<int64_t>(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  // template <int other_dim>
+  // TensorShape(const TensorShape<other_dim> &other) {
+  //   *this = other;
+  // }
+
+  // template <int other_dim>
+  // TensorShape &operator=(const TensorShape<other_dim> &other) {
+  //   data.resize(other.size());
+  //   for (int i = 0; i < size(); i++) {
+  //     data[i] = other[i];
+  //   }
+  //   return *this;
+  // }
+
+  // TensorShape &operator=(TensorShape<-1> &&other) {
+  //   data = std::move(other.data);
+  //   return *this;
+  // }
+
+  // We allow only explicit conversion from dynamic to static dim
+  // template <int other_dim>
+  // explicit operator TensorShape<other_dim>() const;
+  template <int other_dim>
+  TensorShape<other_dim> to_static() const;
+
+  int64_t &operator[](int d) { return shape[d]; }
+  const int64_t &operator[](int d) const { return shape[d]; }
```

Comment:
Returning a const ref to a primitive type can have some undesirable side effects (e.g. can defeat SROA or force object creation to obtain real address).

Reply:
I wanted to be consistent with standard library, which uses const_reference type.